config-model/src/main/java/com/yahoo/vespa/model/application/validation/change/RestartOnDeployForOnnxModelChangesValidator.java

```diff
@@ -32,27 +37,49 @@ public List<ConfigChangeAction> validate(VespaModel currentModel, VespaModel nex
-                if (!currentModels.containsKey(nextModelInfo.modelId())) continue;
-
-                log.log(FINE, "Checking if " + nextModelInfo + " has changed");
-                modelChanged(nextModelInfo, currentModels.get(nextModelInfo.modelId())).ifPresent(change -> {
-                    String message = "Onnx model '%s' has changed (%s), need to restart services in container cluster '%s'"
-                            .formatted(nextModelInfo.modelId(), change, cluster.name());
-                    cluster.onnxModelCostCalculator().setRestartOnDeploy();
-                    actions.add(new VespaRestartAction(cluster.id(), message));
-                });
-            }
+
+            log.log(FINE, "Validating " + cluster + ", current models=" + currentModels + ", next models=" + nextModels);
+            actions.addAll(validateModelChanges(cluster, currentModels, nextModels));
+            actions.addAll(validateSetOfModels(cluster, currentModels, nextModels));
+        }
+        return actions;
+    }
+
+    private List<ConfigChangeAction> validateModelChanges(ApplicationContainerCluster cluster,
+                                                          Map<String, ModelInfo> currentModels,
+                                                          Map<String, ModelInfo> nextModels) {
+        List<ConfigChangeAction> actions = new ArrayList<>();
+        for (var nextModelInfo : nextModels.values()) {
+            if (! currentModels.containsKey(nextModelInfo.modelId())) continue;
+
+            log.log(FINE, "Checking if " + nextModelInfo + " has changed");
+            modelChanged(nextModelInfo, currentModels.get(nextModelInfo.modelId())).ifPresent(change -> {
+                String message = "Onnx model '%s' has changed (%s), need to restart services in %s"
+                        .formatted(nextModelInfo.modelId(), change, cluster);
+                cluster.onnxModelCostCalculator().setRestartOnDeploy();
+                addRestartAction(actions, cluster, message);
+            });
+        }
+        return actions;
+    }
+
+    private List<ConfigChangeAction> validateSetOfModels(ApplicationContainerCluster cluster,
+                                                         Map<String, ModelInfo> currentModels,
+                                                         Map<String, ModelInfo> nextModels) {
+        List<ConfigChangeAction> actions = new ArrayList<>();
+        Set<String> currentModelIds = currentModels.keySet();
+        Set<String> nextModelIds = nextModels.keySet();
+        log.log(FINE, "Checking if model set has changed (%s) -> (%s)".formatted(currentModelIds, nextModelIds));
+        if (! currentModelIds.equals(nextModelIds)) {
+            String message = "Onnx model set has changed from %s to %s, need to restart services in %s"
+                    .formatted(currentModelIds, nextModelIds, cluster);
+            addRestartAction(actions, cluster, message);
```

Comment:
Should `validateSetOfModels` also invoke `setRestartOnDeploy`?

Reply:
Yes, thanks, fixed